### PR TITLE
AP_Camera: runcam backend

### PR DESCRIPTION
**Pull request by andyp1per**
RunCam as a proper camera backend.
Adds a couple of the mavlink info callbacks
Parameter conversion and operation verified on my RunCam split

### CODE REVIEW

**Comment by rmackay9** [comment]
I like where this is going!

**Comment by peterbarker** [comment]
@andyp1per where are you going with this one?

What advantages were you seeing from being a camera backend?

I'd note that the camera library, even in its most compact form is still going to be pretty chunky, and recent commits have shown we don't *really* care about keeping it absolutely tight space-wise.

Definitely not saying it's a bad idea!

**Comment by andyp1per** [comment]
The main thing is to get the mavlink and other support for free. People keep asking for runcam mission support

File: libraries/AP_Camera/AP_Camera.cpp [hunk 1]
```diff
@@ -238,6 +252,15 @@ void AP_Camera::init()
         case CameraType::SCRIPTING:
             _backends[instance] = NEW_NOTHROW AP_Camera_Scripting(*this, _params[instance], instance);
             break;
+#endif
+#if AP_CAMERA_RUNCAM_ENABLED
+        // check for Scripting driver
```
Comment (peterbarker) [inline libraries/AP_Camera/AP_Camera.cpp:257]:
```suggestion
        // check for RunCam driver
```

File: libraries/AP_Camera/AP_Camera.cpp [hunk 2]
```diff
@@ -238,6 +252,15 @@ void AP_Camera::init()
...
+        case CameraType::RUNCAM:
+            _backends[instance] = NEW_NOTHROW AP_RunCam(*this, _params[instance], instance);
+            _backend_var_info[instance] = AP_RunCam::var_info;
+            AP_Param::load_object_from_eeprom(_backends[instance], _backend_var_info[instance]);
+            AP_Param::invalidate_count();
```
Comment (peterbarker) [inline libraries/AP_Camera/AP_Camera.cpp:262]:
```suggestion
```

done in the `load_object_from_eeprom` call

File: libraries/AP_Camera/AP_Camera.cpp [hunk 3]
```diff
@@ -830,7 +853,7 @@ void AP_Camera::convert_params()
 {
     // exit immediately if CAM1_TYPE has already been configured
     if (_params[0].type.configured()) {
-        return;
+        //return;
```
Comment (peterbarker) [inline libraries/AP_Camera/AP_Camera.cpp:856]: This looks strange

File: libraries/AP_Camera/AP_Camera.cpp [hunk 4]
```diff
@@ -847,6 +870,29 @@ void AP_Camera::convert_params()
         // CAM_TRIGG_TYPE was set to Relay, GoPro or Mount
         cam1_type = cam_trigg_type + 1;
     }
+#if AP_CAMERA_RUNCAM_ENABLED
+    // find vehicle's top level key
```
Comment (peterbarker) [inline libraries/AP_Camera/AP_Camera.cpp:874]: Need comments in here with the `PARAMETER_CONVERSION` tag and relevant dates

File: libraries/AP_Camera/AP_Camera_config.h [hunk 5]
```diff
@@ -58,3 +58,7 @@
 #ifndef HAL_RUNCAM_ENABLED
 #define HAL_RUNCAM_ENABLED 1
 #endif
+
+#ifndef AP_CAMERA_RUNCAM_ENABLED
+#define AP_CAMERA_RUNCAM_ENABLED AP_CAMERA_BACKEND_DEFAULT_ENABLED && HAL_RUNCAM_ENABLED
```
Comment (peterbarker) [inline libraries/AP_Camera/AP_Camera_config.h:63]: Any reason to keep `HAL_RUNCAM_ENABLED` around?

File: libraries/AP_Camera/AP_RunCam.cpp [hunk 6]
```diff
@@ -36,7 +36,7 @@ const AP_Param::GroupInfo AP_RunCam::var_info[] = {
     // @DisplayName: RunCam device type
     // @Description: RunCam device type used to determine OSD menu structure and shutter options.
     // @Values: 0:Disabled, 1:RunCam Split Micro/RunCam with UART, 2:RunCam Split, 3:RunCam Split4 4k, 4:RunCam Hybrid/RunCam Thumb Pro, 5:Runcam 2 4k
-    AP_GROUPINFO_FLAGS("TYPE", 1, AP_RunCam, _cam_type, int(DeviceType::Disabled), AP_PARAM_FLAG_ENABLE),
+    AP_GROUPINFO_FLAGS("TYPE", 1, AP_RunCam, _cam_type, int(DeviceType::SplitMicro), AP_PARAM_FLAG_ENABLE),
```
Comment (peterbarker) [inline libraries/AP_Camera/AP_RunCam.cpp:39]: Why the change in default?

File: libraries/AP_Camera/AP_RunCam.cpp [hunk 7]
```diff
@@ -118,13 +118,16 @@ AP_RunCam::Menu AP_RunCam::_menus[RUNCAM_MAX_DEVICE_TYPES] = {
 	{ 6, { 3, 10, 2, 2, 8 }}, // Runcam 2 4K
 };
 
-AP_RunCam::AP_RunCam()
+AP_RunCam::AP_RunCam(AP_Camera &frontend, AP_Camera_Params &params, uint8_t instance)
+    : AP_Camera_Backend(frontend, params, instance)
 {
     AP_Param::setup_object_defaults(this, var_info);
-    if (_singleton != nullptr) {
+    if (_singleton != nullptr && _singleton->_instance == instance) {
```
Comment (peterbarker) [inline libraries/AP_Camera/AP_RunCam.cpp:125]: This looks very bad.  I mean, it's right there in the name...

You probably simply need to remove the singleton from AP_RunCam entirely.

**Review by peterbarker — CHANGES_REQUESTED** [review]

File: libraries/AP_Camera/AP_Camera.cpp (hunk 1 above)
Comment (andyp1per) [reply libraries/AP_Camera/AP_Camera.cpp]: Fixed

File: libraries/AP_Camera/AP_Camera.cpp (hunk 2 above)
Comment (andyp1per) [reply libraries/AP_Camera/AP_Camera.cpp]: Fixed

File: libraries/AP_Camera/AP_Camera.cpp (hunk 3 above)
Comment (andyp1per) [reply libraries/AP_Camera/AP_Camera.cpp]: Fixed

File: libraries/AP_Camera/AP_Camera.cpp (hunk 4 above)
Comment (andyp1per) [reply libraries/AP_Camera/AP_Camera.cpp:952]: Fixed

File: libraries/AP_Camera/AP_Camera_config.h (hunk 5 above)
Comment (andyp1per) [reply libraries/AP_Camera/AP_Camera_config.h:67]: makes it a bigger change to remove it. Would prefer to do a separate PR

File: libraries/AP_Camera/AP_RunCam.cpp (hunk 6 above)
Comment (andyp1per) [reply libraries/AP_Camera/AP_RunCam.cpp:39]: because enabled is now controlled via CAMx_TYPE

**Comment by andyp1per** [comment]
@rmackay9 this is ready to go - any chance of a review?

File: libraries/AP_Camera/AP_Camera.cpp [hunk 8]
```diff
@@ -41,10 +43,24 @@ const AP_Param::GroupInfo AP_Camera::var_info[] = {
     // @Path: AP_Camera_Params.cpp
     AP_SUBGROUPINFO(_params[1], "2", 13, AP_Camera, AP_Camera_Params),
 #endif
+#if AP_CAMERA_RUNCAM_ENABLED
+    // @Group: 1_RC_
+    // @Path: AP_RunCam.cpp
+    AP_SUBGROUPVARPTR(_backends[0], "1_RC_", 14, AP_Camera, _backend_var_info[0]),
```
Comment (rmackay9) [inline libraries/AP_Camera/AP_Camera.cpp:49]: Is this how we normally handle extra parameters required by a particular backend?  I wonder if @peterbarker has advice on this

File: libraries/AP_Camera/AP_RunCam.cpp (hunk 6 above)
Comment (rmackay9) [inline libraries/AP_Camera/AP_RunCam.cpp:39]: I'm not suggesting we change it but we could name this parameter "model" to avoid confusion with CAMx_TYPE.  Another option would be to have different entries in the CAM1_TYPE for each model and then pass that into the constructor and then remove the 1_RC_TYPE parameter completely

File: libraries/AP_Camera/AP_Camera.cpp (hunk 8 above)
Comment (peterbarker) [reply libraries/AP_Camera/AP_Camera.cpp:49]: Yeah, this is pretty typical of the way we do this.

File: libraries/AP_Camera/AP_Camera.cpp [hunk 9]
```diff
@@ -193,6 +209,8 @@ void AP_Camera::init()
         return;
     }
 
+    memset(_backends, 0, sizeof(void*)*AP_CAMERA_MAX_INSTANCES);
```
Comment (peterbarker) [inline libraries/AP_Camera/AP_Camera.cpp:212]: Why is this being done?!
```suggestion
```

File: libraries/AP_Camera/AP_Camera.cpp [hunk 10]
```diff
@@ -919,6 +952,42 @@ void AP_Camera::convert_params()
...
+        rc_type = int8_t(CameraType::RUNCAM);
+        _backends[1] = NEW_NOTHROW AP_RunCam(*this, _params[1], 1, _runcam_instances);
+        _backend_var_info[1] = AP_RunCam::var_info;
+        AP_Param::convert_class(k_param_vehicle_key, &_backends[1], _backend_var_info[1], 1, false);
+        AP_Param::invalidate_count();
+        _runcam_instances++;
+    }
+
+    _params[1].type.set_and_save(rc_type);
+#endif
```
Comment (peterbarker) [inline libraries/AP_Camera/AP_Camera.cpp:989]:
```suggestion
#endif  // AP_CAMERA_RUNCAM_ENABLED
```

File: libraries/AP_Camera/AP_RunCam.cpp [hunk 11]
```diff
@@ -118,13 +118,24 @@ AP_RunCam::Menu AP_RunCam::_menus[RUNCAM_MAX_DEVICE_TYPES] = {
 	{ 6, { 3, 10, 2, 2, 8 }}, // Runcam 2 4K
 };
 
-AP_RunCam::AP_RunCam()
+const char* AP_RunCam::_models[RUNCAM_MAX_DEVICE_TYPES] = {
```
Comment (peterbarker) [inline libraries/AP_Camera/AP_RunCam.cpp:121]:
```suggestion
static const char* AP_RunCam::_models[RUNCAM_MAX_DEVICE_TYPES] {
```

File: libraries/AP_Camera/AP_RunCam.cpp [hunk 12]
```diff
@@ -118,13 +118,24 @@ AP_RunCam::Menu AP_RunCam::_menus[RUNCAM_MAX_DEVICE_TYPES] = {
...
+};
+
+AP_RunCam::AP_RunCam(AP_Camera &frontend, AP_Camera_Params &params, uint8_t instance, uint8_t runcam_instance)
+    : AP_Camera_Backend(frontend, params, instance), _runcam_instance(runcam_instance)
 {
     AP_Param::setup_object_defaults(this, var_info);
-    if (_singleton != nullptr) {
-        AP_HAL::panic("AP_RunCam must be singleton");
+    if (_singleton != nullptr && _singleton->_instance == instance) {
+        AP_HAL::panic("AP_RunCam instance must be a singleton %u\n", instance);
```
Comment (peterbarker) [inline libraries/AP_Camera/AP_RunCam.cpp:134]:
```suggestion
        AP_HAL::panic("AP_RunCam instance must be a singleton %u", instance);
```

if there are places that need to add a line ending then patch those places.

File: libraries/AP_Camera/AP_RunCam.cpp [hunk 13]
```diff
@@ -1058,8 +1069,86 @@ uint8_t AP_RunCam::Request::get_expected_response_length(const Command command)
...
+void AP_RunCam::send_camera_information(mavlink_channel_t chan) const
+{
+    // exit immediately if not initialised
+    if (!camera_ready()) {
+        return;
+    }
+
+    static const uint8_t vendor_name[32] = "RunCam";
+    uint8_t model_name[32] {};
+    strncpy((char *)model_name, _models[_cam_type.get()-1], MIN(sizeof(model_name), sizeof(_models[_cam_type.get()-1])));
```
Comment (peterbarker) [inline libraries/AP_Camera/AP_RunCam.cpp:1100]: This looks very bad - indexing an array from a user-set parameter looks like a really bad idea.

What happens if you change CAM1_TYPE to 0 after it has detected a runcam?

File: libraries/AP_Camera/AP_RunCam.cpp [hunk 14]
```diff
@@ -1058,8 +1069,86 @@ uint8_t AP_RunCam::Request::get_expected_response_length(const Command command)
...
+    // send CAMERA_INFORMATION message
+    mavlink_msg_camera_information_send(
+        chan,
+        AP_HAL::millis(),       // time_boot_ms
+        vendor_name,            // vendor_name uint8_t[32]
+        model_name,             // model_name uint8_t[32]
+        0,                      // firmware version uint32_t
+        NaNf,                   // sensor_size_h float (mm)
+        NaNf,                   // sensor_size_v float (mm)
+        0,                      // sensor_size_v float (mm)
```
Comment (peterbarker) [inline libraries/AP_Camera/AP_RunCam.cpp:1123]: bad comment?

File: libraries/AP_HAL_ChibiOS/hwdef/include/minimize_common.inc [hunk 15]
```diff
@@ -48,6 +48,7 @@ define AP_WINCH_ENABLED 0
 define AP_CAMERA_BACKEND_DEFAULT_ENABLED 0
 define AP_CAMERA_RELAY_ENABLED AP_CAMERA_ENABLED
 define AP_CAMERA_SERVO_ENABLED AP_CAMERA_ENABLED
+define AP_CAMERA_RUNCAM_ENABLED AP_CAMERA_ENABLED
```
Comment (peterbarker) [inline libraries/AP_HAL_ChibiOS/hwdef/include/minimize_common.inc:51]: This doesn't look right.  We shouldn't be adding things to minimized boards like this - we only have these other backends in here so we don't break people's builds!

**Review by peterbarker — CHANGES_REQUESTED** [review]

File: libraries/AP_Camera/AP_RunCam.cpp (hunk 13 above)
Comment (peterbarker) [reply libraries/AP_Camera/AP_RunCam.cpp:1100]: Maybe check the configured type is the same as the allocated type in here?

File: libraries/AP_Camera/AP_RunCam.cpp (hunk 6 above)
Comment (andyp1per) [reply libraries/AP_Camera/AP_RunCam.cpp]: changed to model

File: libraries/AP_Camera/AP_Camera.cpp (hunk 9 above)
Comment (andyp1per) [reply libraries/AP_Camera/AP_Camera.cpp:212]: Because the array is not zeroed and we check whether the backend has been allocated.

File: libraries/AP_Camera/AP_Camera.cpp (hunk 10 above)
Comment (andyp1per) [reply libraries/AP_Camera/AP_Camera.cpp]: Fixed

File: libraries/AP_Camera/AP_RunCam.cpp (hunk 11 above)
Comment (andyp1per) [reply libraries/AP_Camera/AP_RunCam.cpp:121]: naughty, naughty - asking me to change to invalid C++. This is already declared static in the header (where it should be).

File: libraries/AP_Camera/AP_RunCam.cpp (hunk 12 above)
Comment (andyp1per) [reply libraries/AP_Camera/AP_RunCam.cpp:134]: I'm not going to patch all uses of panic() in this PR. There are other places where linefeed is added to panic lines and it looks rubbish without it.

File: libraries/AP_Camera/AP_RunCam.cpp (hunk 14 above)
Comment (andyp1per) [reply libraries/AP_Camera/AP_RunCam.cpp:1123]: The comment is correct - we are just not filling in these fields

File: libraries/AP_HAL_ChibiOS/hwdef/include/minimize_common.inc (hunk 15 above)
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/hwdef/include/minimize_common.inc:51]: RunCam is already enabled in the minimized builds, this just makes sure it continues to be so

**Review by rmackay9 — APPROVED** [review]
I guess we could rename the class to be AP_Camera_RunCam but I don't think we should slow-walk this any more.  It's an improvement and it's the beginning of the 4.7 dev cycle so looks OK to me, thanks!

File: libraries/AP_Camera/AP_Camera.cpp (hunk 9 above)
Comment (tridge) [reply libraries/AP_Camera/AP_Camera.cpp:212]: that is a surprise, as anything not stack allocated should start as zero. It would be worth trying to find out why it is non-zero, we rely on that pretty heavily

**Comment by rmackay9** [comment]
BTW, my RunCam arrived today in the mail so I'm looking forward to testing this (regardless of whether it's already merged to master or still in PR form)

File: libraries/AP_Camera/AP_Camera.cpp (hunk 9 above)
Comment (andyp1per) [reply libraries/AP_Camera/AP_Camera.cpp]: Ok I removed it and it still seems to work, so maybe this was only needed when my code was broken 😄

File: Tools/scripts/build_options.py [hunk 16]
```diff
@@ -146,7 +146,7 @@ def config_option(self):
     Feature('Camera', 'Camera_ThermalRange', 'AP_CAMERA_SEND_THERMAL_RANGE_ENABLED', 'Enable GCS camera thermal range', 0, 'Camera,MOUNT'),  # noqa: E501
     Feature('Camera', 'Camera_Info_From_Script', 'AP_CAMERA_INFO_FROM_SCRIPT_ENABLED', 'Enable Camera information messages via Lua script', 0, 'Camera,SCRIPTING'), # noqa
 
-    Feature('Camera', 'RUNCAM', 'HAL_RUNCAM_ENABLED', 'Enable RunCam control', 0, None),
+    Feature('Camera', 'RUNCAM', 'AP_CAMERA_RUNCAM_ENABLED', 'Enable RunCam Control', 0, None),
```
Comment (peterbarker) [inline Tools/scripts/build_options.py:149]:
```suggestion
    Feature('Camera', 'RUNCAM', 'AP_CAMERA_RUNCAM_ENABLED', 'Enable RunCam control', 0, None),
```

Your patch here is reverting a chnage @Hwurzburg made to make all of our strings consistent in the custom build server.

File: libraries/AP_Camera/AP_RunCam.cpp (hunk 7 above)
Comment (peterbarker) [reply libraries/AP_Camera/AP_RunCam.cpp:133]: Right, but the thing isn't a singleton.  It's just the first instance.  This means we've got a significantly different meaning for the word "singleton" within the RunCam library to everywhere else in ArduPilot.

This can be done better in several different ways.  You can create specially-named methods on AP::runcam() - eg. `AP::camera().runcam_start_recording()`, and we could also look at polling the aux-switch state in the backend.  Could also just rename singleton to `_first_instance`.

... but I really can't ask you to do that in this PR, so I think we just put up with this pretty nasty discrepancy.

**Review by peterbarker — CHANGES_REQUESTED** [review]

File: libraries/AP_Camera/AP_RunCam.cpp (hunk 12 above)
Comment (andyp1per) [reply libraries/AP_Camera/AP_RunCam.cpp:134]: Ok fixed

File: libraries/AP_Camera/AP_RunCam.cpp (hunk 13 above)
Comment (andyp1per) [reply libraries/AP_Camera/AP_RunCam.cpp:1100]: No, no - check the if at the top of the function

File: libraries/AP_Camera/AP_RunCam.cpp [hunk 17]
```diff
@@ -1058,8 +1069,86 @@ uint8_t AP_RunCam::Request::get_expected_response_length(const Command command)
...
```
[the remark this answers is not in the document]
Comment (andyp1per) [reply libraries/AP_Camera/AP_RunCam.cpp:1123]: Ok this bug is actually in the mount library as well, I have changed the commenst and value sto be correct

File: Tools/scripts/build_options.py (hunk 16 above)
Comment (andyp1per) [reply Tools/scripts/build_options.py]: Fixed

File: libraries/AP_Camera/AP_RunCam.cpp (hunk 7 above)
Comment (andyp1per) [reply libraries/AP_Camera/AP_RunCam.cpp:133]: Fair comment

File: libraries/AP_Camera/AP_RunCam.cpp (hunk 13 above)
Comment (peterbarker) [reply libraries/AP_Camera/AP_RunCam.cpp:1102]: Looks right now

File: libraries/AP_Camera/AP_RunCam.cpp [hunk 18]
```diff
@@ -1058,8 +1069,88 @@ uint8_t AP_RunCam::Request::get_expected_response_length(const Command command)
...
+    }
+
+    // send CAMERA_INFORMATION message
+    mavlink_msg_camera_information_send(
+        chan,
+        AP_HAL::millis(),       // time_boot_ms
+        vendor_name,            // vendor_name uint8_t[32]
+        model_name,             // model_name uint8_t[32]
+        0,                      // firmware version uint32_t
+        0,                      // focal_length float (mm)
```
Comment (peterbarker) [inline libraries/AP_Camera/AP_RunCam.cpp:1123]:
```suggestion
        NaNf,                      // focal_length float (mm)
```

File: libraries/AP_Camera/AP_RunCam.cpp (hunk 13 above)
Comment (andyp1per) [reply libraries/AP_Camera/AP_RunCam.cpp:1102]: OK fixed

**Review by peterbarker — CHANGES_REQUESTED** [review]

**Review by peterbarker — APPROVED** [review]
